File: TJUBBS/BBCodeParser.js
// aceptable BBcode tags, optionally prefixed with a slash
var tagname_re = /^\/?(?:b|i|u|table|tr|td|font|\*|pre|center|left|right|samp|code|colou?r|backcolou?r|size|noparse|url|link|s|q|(block)?quote|img|list|float|align|li|attimg)$/i;
// color names or hex color
var color_re = /^(:?[A-Za-z]+|#(?:[0-9a-f]{3})?[0-9a-f]{3}|[A-Za-z]+\([0-9,%.]+\))$/i;
// numbers
var number_re = /^[\\.0-9]{1,8}$/i;
// reserved, unreserved, escaped and alpha-numeric [RFC2396]
var uri_re = /^[-;/?:@&=+$,_.!~*'()%0-9a-z]{1,512}$/i;
// main regular expression: CRLF, [tag=option], [tag="option"] [tag] or [/tag]
var postfmt_re = /([\r\n])|(?:\[([a-z@#]{1,16})(?:=([^\x00-\x1F<>[\]]{1,256}))?\])|(?:\[\/([a-z]{1,16})\])/ig;

function BBCode(post, cb) {
  // var attimg = [];
  var opentags = []; // open tag stack
  var crlf2br = true; // convert CRLF to <br>?
  var noparse = false; // ignore BBCode tags?
  var urlstart = -1; // beginning of the URL if zero or greater (ignored if -1)

  // stack frame object
  function taginfo_t(bbtag, etag) {
    return {
      bbtag: bbtag,
      etag: etag
    };
  }

  // check if it's a valid BBCode tag
  function isValidTag(str) {
    if (!str || !str.length)
      return false;
    return tagname_re.test(str);
  }

  /*
   m1 - CR or LF
   m2 - the tag of the [tag=option] expression
   m3 - the option of the [tag=option] expression
   m4 - the end tag of the [/tag] expression
   */
  function textToHtmlCB(mstr, m1, m2, m3, m4, offset, string) {
    // CR LF sequences
    if (m1 && m1.length) {
      if (!crlf2br)
        return mstr;
      switch (m1) {
        case '\r':
          return "";
        case '\n':
          return "<br>";
        default:
          break;
      }
    }
    // handle start tags
    if (isValidTag(m2)) {
      var m2l = m2.toLowerCase();
      // if in the noparse state, just echo the tag
      if (noparse)
        return "[" + m2 + "]";
      // ignore any tags if there's an open option-less [url] tag
      if (opentags.length && (opentags[opentags.length - 1].bbtag === "url" || opentags[opentags.length - 1].bbtag === "link") && urlstart >= 0)
        return "[" + m2 + "]";
      switch (m2l) {
        case "code":
          opentags.push(taginfo_t(m2l, "</code></pre>"));
          crlf2br = false;
          return "<pre><code>";
        case "pre":
          opentags.push(taginfo_t(m2l, "</pre>"));
          crlf2br = false;
          return "<pre>";
        case "center":
        case "left":
        case "right":
          opentags.push(taginfo_t(m2l, "</div>"));
          return '<div style="text-align:'+m2l+'">';
        case "align":
          opentags.push(taginfo_t(m2l, "</div>"));
          if (m3 === 'left' || m3 === 'center' || m3 === 'right') {
            return '<div style="text-align:' + m3 + '">';
          } else {
            return '<div>';
          }
        case "float":
          opentags.push(taginfo_t(m2l, "</div>"));
          if (m3 === 'left' || m3 === 'right') {
            return '<div style="float:' + m3 + '">';
          } else {
            return '<div>';
          }
        case "color":
        case "colour":
          if (!m3 || !color_re.test(m3))
            m3 = "inherit";
          opentags.push(taginfo_t(m2l, "</span>"));
          return "<span style=\"color: " + m3 + "\">";
        case "backcolor":
        case "backcolour":
          if (!m3 || !color_re.test(m3))
            m3 = "inherit";
          opentags.push(taginfo_t(m2l, "</span>"));
          return "<span style=\"background-color: " + m3 + "\">";
        case "font":
          opentags.push(taginfo_t(m2l, "</span>"));
          if (!m3)
            return '<span>';
          else
            return "<span style=\"font-family: " + m3.replace(/"/g, '"') + "\">";
        case "size":
          if (!m3 || !number_re.test(m3)) {
            m3 = '1em';
          } else {
            m3 += 'px';
          }
          opentags.push(taginfo_t(m2l, "</span>"));
          return "<span style=\"font-size: " + m3 + "\">";
        case "s":
          opentags.push(taginfo_t(m2l, "</span>"));
          return "<span style=\"text-decoration: line-through\">";
        case "noparse":
          noparse = true;
          return "";
        case "link":
        case "url":
          opentags.push(taginfo_t(m2l, "</a>"));
          if (m3 && uri_re.test(m3)) {
            urlstart = -1;
            return "<a target=\"_blank\" href=\"" + m3 + "\">";
          }
          urlstart = mstr.length + offset;
          return "<a target=\"_blank\" href=\"";
        case "img":
          opentags.push(taginfo_t(m2l, "\" />"));
          if (m3) {
            var size = m3.split(',');
            return '<' + m2l + ' width="'+size[0]+'" height="'+size[1]+'" src="';
          }
          return "<" + m2l + " src=\"";
        case "q":
        case "quote":
        case "blockquote":
          var qTag = (m2l === "q") ? "q" : "blockquote";
          opentags.push(taginfo_t(m2l, "</" + qTag + ">"));
          return m3 && m3.length && uri_re.test(m3) ? "<" + qTag + " cite=\"" + m3 + "\">" : "<" + qTag + ">";
        case "list":
          var lTtag = m3 && m3 === '1' ? 'ol' : 'ul';
          crlf2br = false;
          opentags.push(taginfo_t('list', '</'+lTtag+'>'));
          return '<'+lTtag+'>';
        case "*":
          opentags.push(taginfo_t('*', '</li>'));
          return '<li>';
        case "b":
          opentags.push(taginfo_t('b', '</strong>'));
          return '<strong>';
        case "i":
          opentags.push(taginfo_t('i', '</em>'));
          return '<em>';
        default:
          // [samp] [u] [table] [tr] [td]
          opentags.push(taginfo_t(m2l, "</" + m2l + ">"));
          return "<" + m2l + ">";
      }
    }
    //
    // process end tags
    //
    if (isValidTag(m4)) {
      var m4l = m4.toLowerCase();
      if (noparse) {
        // if it's the closing noparse tag, flip the noparse state
        if (m4 === "noparse") {
          noparse = false;
          return "";
        }
        // otherwise just output the original text
        return "[/" + m4 + "]";
      }
      // highlight mismatched end tags
      if (!opentags.length || opentags[opentags.length - 1].bbtag !== m4l)
        return "<span style=\"color: red\">[/" + m4 + "]</span>";
      if (m4l === "url" || m4l === "link") {
        // if there was no option, use the content of the [url] tag
        if (urlstart > 0)
          return "\">" + string.substr(urlstart, offset - urlstart) + opentags.pop().etag;
        // otherwise just close the tag
        return opentags.pop().etag;
      }
      else if (m4l === "code" || m4l === "pre" || m4l === "list")
        crlf2br = true;
      else if (m4l === "*")
        crlf2br = false;
      // other tags require no special processing, just output the end tag
      var end = opentags.pop().etag;
      return end;
    }
    return mstr;
  }

  // actual parsing can begin
  var result = '', endtags;
  // convert CRLF to <br> by default
  crlf2br = true;
  // create a new array for open tags
  if (opentags == null || opentags.length)
    opentags = new Array(0);
  // run the text through main regular expression matcher
  if (post) {
    post = post.trim().replace(/&/g, '&amp;').replace(/</g, '&lt;')
      .replace(/>/g, '&gt;').replace(/"/g, '&quot;');

    post = post.replace(/\[\*\](.*)/g, function (m0, m1) {
      return '[li]'+m1+'[/li]';
    });

    post = post.replace(/\[attach(?:=(\d+))?\](.*?)\[\/attach\]/g, function (m0, m1, m2) {
      if (m1) {
        return `<a href="https://bbs.twtstudio.com/api/attach/${m1}?name=${encodeURIComponent(m2)}" target="_blank">附件: ${m2}</a>`;
      } else {
        return `<a href="https://bbs.twtstudio.com/api/attach/${m1}" target="_blank">附件: <i>(未命名: ${m1})</i></a>`;
      }
    });

    post = post.replace(/\[attimg\](.*?)\[\/attimg\]/g, function (m0, m1) {
      return `<img src="https://bbs.twtstudio.com/api/img/${m1}"></a>`;
    });
    // @todo: 附件使用情况统计
    /*
    post = _post.replace(/\[attimg\][/attimg]/g, function (m0, m1) {
      return '['+m1+'li]';
    });
    */
    result = post.replace(postfmt_re, textToHtmlCB);
    // reset noparse, if it was unbalanced
    if (noparse)
      noparse = false;
    // if there are any unbalanced tags, make sure to close them
    if (opentags.length) {
      endtags = '';
      // if there's an open [url] at the top, close it
      if (opentags[opentags.length - 1].bbtag === "url" || opentags[opentags.length - 1].bbtag === "link") {
        opentags.pop();
        endtags += "\">" + post.substr(urlstart, post.length - urlstart) + "</a>";
      }
      // close remaining open tags
      while (opentags.length)
        endtags += opentags.pop().etag;
    }
  }

  result = endtags ? result + endtags : result;
  // & has been escaped to &amp; before processing, 
  return result.replace(/&amp;#91;/g, '[').replace(/&amp;#93;/g, ']');
}

File: TJUBBS/BBCodeParser.swift



//
//  BBCodeParser.swift
//  TJUBBS
//
//  Created by Halcao on 2017/6/2.
//  Copyright © 2017年 twtstudio. All rights reserved.
//

import UIKit

let base = "https://bbs.tju.edu.cn"
struct BBCodeParser {
    private static var map: [String : String] {
        get {
            var map = [String : String]()
            map = ["&amp;rsquo;":"'",
                   "&lt":"<",
                   "&gt":">",
                   "&rsquo;":"'",
                   "\n":"<br />",
                   
                   /* lowercase */
                
                "\\[b\\](.+?)\\[/b\\]":"<strong>$1</strong>",
                "\\[i\\](.+?)\\[/i\\]":"<i>$1</i>",
                "\\[u\\](.+?)\\[/u\\]":"<u>$1</u>",
                "\\[s\\](.+?)\\[/s\\]":"<s>$1</s>",
                "\\[h1\\](.+?)\\[/h1\\]":"<h1>$1</h1>",
                "\\[h2\\](.+?)\\[/h2\\]":"<h2>$1</h2>",
                "\\[h3\\](.+?)\\[/h3\\]":"<h3>$1</h3>",
                "\\[h4\\](.+?)\\[/h4\\]":"<h4>$1</h4>",
                "\\[h5\\](.+?)\\[/h5\\]":"<h5>$1</h5>",
                "\\[h6\\](.+?)\\[/h6\\]":"<h6>$1</h6>",
                "\\[quote\\]([\\s\\S]+?)\\[/quote\\]":"<blockquote>$1</blockquote>",
                "\\[quote=(.+?)\\](.+?)\\[/quote\\]":"<blockquote>$2</blockquote>",
                "\\[p\\](.+?)\\[/p\\]":"<p>$1</p>",
                "\\[p=(.+?),(.+?)\\](.+?)\\[/p\\]":"<p style=\"text-indent:$1px;line-height:$2%;\">$3</p>",
                "\\[center\\](.+?)\\[/center\\]":"<div align=\"center\">$1",
                "\\[align=(.+?)\\](.+?)\\[/align\\]":"<div align=\"$1\">$2",
                "\\[color=(.+?)\\](.+?)\\[/color\\]":"<font color='$1'>$2</font>",
                "\\[size=(.+?)\\](.+?)\\[/size\\]":"<font size=\"$1\">$2</span>",
//                "\\[img\\](.+?)\\[/img\\]":"<img src=\"\(base)$1\" width=\"100\" height=\"100\"/>",
                "\\[img\\](.+?)\\[/img\\]":"<img src=\"\(base)$1\"/>",
                
                //乔成骚改的
                "\\[attimg\\](.+?)\\[/attimg\\]":"<img src=\"\(base)/api/attach/$1\" />",
                
                "\\[img=(.+?),(.+?)\\](.+?)\\[/img\\]":"<img width=\"$1\" height=\"$2\" src=\"\(base)$3\" />",
                "\\[email\\](.+?)\\[/email\\]":"<a href=\"mailto:$1\">$1</a>",
                "\\[email=(.+?)\\](.+?)\\[/email\\]":"<a href=\"mailto:$1\">$2</a>",
                "\\[url\\](.+?)\\[/url\\]":"<a href=\"$1\">$1</a>",
                "\\[url=(.+?)\\](.+?)\\[/url\\]":"<a href=\"$1\">$2</a>",
                "\\[youtube\\](.+?)\\[/youtube\\]":"<object width=\"640\" height=\"380\"><param name=\"movie\" value=\"http://www.youtube.com/v/$1\"></param><embed src=\"http://www.youtube.com/v/$1\" type=\"application/x-shockwave-flash\" width=\"640\" height=\"380\"></embed></object>",
                "\\[video\\](.+?)\\[/video\\]":"<video src=\"$1\" />",
                
                
                "\\[li\\](.+?)\\[/il\\]":"<li>$1</li>",
                "\\[ol\\](.+?)\\[/ol\\]":"<ol>$1</ol>",
                "\\[ul\\](.+?)\\[/ul\\]":"<ul>$1</ul>",
                
                "\\[list\\](.+?)\\[/list\\]":"<ul>$1</ul>",
                "\\[code\\](.+?)\\[/code\\]":"<code>$1</code>",
                //        "\\[center\\](.+?)\\[/center\\]":"<br /><center>$1</center>"]
                
                /* UPPERCASE */
                
                "\\[B\\](.+?)\\[/B\\]":"<STRONG>$1</STRONG>",
                "\\[I\\](.+?)\\[/I\\]":"<I>$1</I>",
                "\\[U\\](.+?)\\[/U\\]":"<U>$1</U>",
                "\\[S\\](.+?)\\[/S\\]":"<S>$1</S>",
                "\\[H1\\](.+?)\\[/H1\\]":"<H1>$1</H1>",
                "\\[H2\\](.+?)\\[/H2\\]":"<H2>$1</H2>",
                "\\[H3\\](.+?)\\[/H3\\]":"<H3>$1</H3>",
                "\\[H4\\](.+?)\\[/H4\\]":"<H4>$1</H4>",
                "\\[H5\\](.+?)\\[/H5\\]":"<H5>$1</H5>",
                "\\[H6\\](.+?)\\[/H6\\]":"<H6>$1</H6>",
                "\\[QUOTE\\](.+?)\\[/QUOTE\\]":"<BLOCKQUOTE>$1</BLOCKQUOTE>",
                "\\[QUOTE=(.+?)\\](.+?)\\[/QUOTE\\]":"<BLOCKQUOTE>$2</BLOCKQUOTE>",
                "\\[P\\](.+?)\\[/P\\]":"<P>$1</P>",
                "\\[P=(.+?),(.+?)\\](.+?)\\[/P\\]":"<P STYLE=\"TEXT-INDENT:$1PX;LINE-HEIGHT:$2%;\">$3</P>",
                "\\[CENTER\\](.+?)\\[/CENTER\\]":"<DIV ALIGN=\"CENTER\">$1",
                "\\[ALIGN=(.+?)\\](.+?)\\[/ALIGN\\]":"<DIV ALIGN=\"$1\">$2",
                "\\[COLOR=(.+?)\\](.+?)\\[/COLOR\\]":"<FONT COLOR='$1'>$2</FONT>",
                "\\[SIZE=(.+?)\\](.+?)\\[/SIZE\\]":"<SPAN STYLE=\"FONT-SIZE:$1;\">$2</SPAN>",
//                "\\[IMG\\](.+?)\\[/IMG\\]":"<IMG SRC=\"\(base)$1\" width=\"100\" height=\"100\"/>",
                "\\[IMG\\](.+?)\\[/IMG\\]":"<IMG SRC=\"\(base)$1\"/>",
                "\\[IMG=(.+?),(.+?)\\](.+?)\\[/IMG\\]":"<IMG WIDTH=\"$1\" HEIGHT=\"$2\" SRC=\"\(base)$3\" />",
                "\\[EMAIL\\](.+?)\\[/EMAIL\\]":"<A HREF=\"MAILTO:$1\">$1</A>",
                "\\[EMAIL=(.+?)\\](.+?)\\[/EMAIL\\]":"<A HREF=\"MAILTO:$1\">$2</A>",
                "\\[URL\\](.+?)\\[/URL\\]":"<A HREF=\"$1\">$1</A>",
                "\\[URL=(.+?)\\](.+?)\\[/URL\\]":"<A HREF=\"$1\">$2</A>",
                "\\[YOUTUBE\\](.+?)\\[/YOUTUBE\\]":"<OBJECT WIDTH=\"640\" HEIGHT=\"380\"><PARAM NAME=\"MOVIE\" VALUE=\"HTTP://WWW.YOUTUBE.COM/V/$1\"></PARAM><EMBED SRC=\"HTTP://WWW.YOUTUBE.COM/V/$1\" TYPE=\"APPLICATION/X-SHOCKWAVE-FLASH\" WIDTH=\"640\" HEIGHT=\"380\"></EMBED></OBJECT>",
                "\\[VIDEO\\](.+?)\\[/VIDEO\\]":"<VIDEO SRC=\"$1\" />",
                "\\[LIST\\](.+?)\\[/LIST\\]":"<UL>$1</UL>",
                "\\[LI\\](.+?)\\[/IL\\]":"<LI>$1</LI>",
                "\\[OL\\](.+?)\\[/OL\\]":"<OL>$1</OL>",
                "\\[UL\\](.+?)\\[/UL\\]":"<UL>$1</UL>",
                "\\[CODE\\](.+?)\\[/CODE\\]":"<CODE>$1</CODE>"]
            //        "\\[CENTER\\](.+?)\\[/CENTER\\]":"<br /><CENTER>$1</CENTER>"]
            return map
        }
    }
    
    static func parse(string: String) -> String {
        guard string != "" else {
            return ""
        }
        var result = string
        for pattern in Array(map.keys) {
            let regex = try! NSRegularExpression(pattern: pattern, options: NSRegularExpression.Options.caseInsensitive)
            let range = NSMakeRange(0, result.utf16.count)
            result = regex.stringByReplacingMatches(in: result, options: [], range: range, withTemplate: map[pattern]!)
        }
        return result
    }
}
